### ISSUE TITLE
After recent update, '\' , '|' are changed to _ and dynamic prompts don't work

### ISSUE THREAD

**Issue by thecakeisal1e** [adieyal/sd-dynamic-prompts#280]
> Error running process: .\stable-diffusion-webui\extensions\sd-dynamic-prompts\scripts\dynamic_prompting.py
Traceback (most recent call last):
  File ".\stable-diffusion-webui\modules\scripts.py", line 386, in process
    script.process(p, *script_args)
  File ".\stable-diffusion-webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py", line 412, in process
    all_prompts, all_negative_prompts = generate_prompts(
  File ".\stable-diffusion-webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py", line 73, in generate_prompts
    all_prompts = prompt_generator.generate(prompt, num_prompts)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\generators\magicprompt.py", line 143, in generate
    prompts = self._prompt_generator.generate(*args, **kwargs)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\generators\randomprompt.py", line 47, in generate
    return list(self._sampler.generate_prompts(template, num_images))
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\random.py", line 77, in generator_from_command
    yield command.separator.join(next(subgen) for subgen in sub_generators)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\random.py", line 77, in <genexpr>
    yield command.separator.join(next(subgen) for subgen in sub_generators)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\random.py", line 89, in generator_from_command
    yield from self.generate_prompts(value, 1)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\random.py", line 89, in generator_from_command
    yield from self.generate_prompts(value, 1)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\base.py", line 50, in generate_prompts
    command = parse(prompt, parser_config=self._parser_config)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\parser\parse.py", line 257, in parse
    tokens = create_parser(parser_config=parser_config).parse_string(
  File ".\stable-diffusion-webui\venv\lib\site-packages\pyparsing\core.py", line 1141, in parse_string
    raise exc.with_traceback(None)
pyparsing.exceptions.ParseException: Expected end of text, found '{'  (at char 28), (line:1, col:29)

**Comment by adieyal** [comment]
Could you provide a prompt that I can use for testing?

I just tried `A __colours__ {triangle|square|circle}` which works as expected.

**Comment by adieyal** [comment]
@thecakeisal1e  Is this still an issue for you?

**Comment by thecakeisal1e** [comment]
Ummm, I'd have to upload all my wildcards.
My high-level prompt is something like this:
__Quality__, __Perspectives__, __Framing__, __Color_Palettes__, {2$$__Image_Composition__}, __Backgrounds__, __body/Face__, __body/Posture_Body__, __body/Gazes__, __body/Arm_Posture__

and the most complex wildcard looks like this:
__Backgrounds__
indoors, classroom, {1-2$$ and $${1-2$$ and $${window|door|doorway|hallway}}|{1-2$$ and $${table|desk|chair|bookshelf}}|{tile_floor|wooden_floor}}, __Time_of_Day__
indoors, classroom, {1-2$$ and $${1-2$$ and $${window|door|doorway|hallway}}|{1-2$$ and $${table|desk|chair|bookshelf}}|{tile_floor|wooden_floor}}

this set of wildcards and prompts was working until the most recent update and the problem persists

**Comment by thecakeisal1e** [comment]
Yeah, it's still a problem. My prompts are all based on nested wildcards but each one of them comes out as just the names of the wildcards and, if it references a subdirectory, the / is changed to a _ and the | in a {  |  } structure is changed to a _.

**Comment by thecakeisal1e** [comment]
This is the companion error. Seems similar to the first.

> Error running process: .\stable-diffusion-webui\extensions\sd-dynamic-prompts\scripts\dynamic_prompting.py
Traceback (most recent call last):
  File ".\stable-diffusion-webui\modules\scripts.py", line 386, in process
    script.process(p, *script_args)
  File ".\stable-diffusion-webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py", line 413, in process
    all_prompts, all_negative_prompts = generate_prompts(
  File ".\stable-diffusion-webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py", line 73, in generate_prompts
    all_prompts = prompt_generator.generate(prompt, num_prompts) or [""]
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\generators\randomprompt.py", line 46, in generate
    return list(self._context.sample_prompts((template or ""), num_images))
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\sampling_context.py", line 106, in <genexpr>
    gen = (squash_whitespace(p) for p in gen)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\base.py", line 64, in _get_sequence
    yield rotate_and_join(sub_generators, separator=command.separator)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\utils.py", line 43, in rotate_and_join
    return separator.join(rotate_all(generators))
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\utils.py", line 35, in rotate_all
    return [next(gen) for gen in generators]
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\utils.py", line 35, in <listcomp>
    return [next(gen) for gen in generators]
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\random.py", line 98, in _get_wildcard
    yield from context.sample_prompts(value, 1)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\sampling_context.py", line 106, in <genexpr>
    gen = (squash_whitespace(p) for p in gen)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\samplers\random.py", line 98, in _get_wildcard
    yield from context.sample_prompts(value, 1)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\sampling_context.py", line 97, in sample_prompts
    command = parse(prompt, parser_config=self.parser_config)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\parser\parse.py", line 302, in parse
    tokens = create_parser(parser_config=parser_config).parse_string(
  File ".\stable-diffusion-webui\venv\lib\site-packages\pyparsing\core.py", line 1141, in parse_string
    raise exc.with_traceback(None)
pyparsing.exceptions.ParseException: Expected end of text, found '{'  (at char 49), (line:1, col:50)

I've went through and tried toggling any options available to me and building up my prompts from different wildcard combinations. It seems like it might be involved with wildcards that contain references to other wildcards or with options of wildcards in them (ie __Backgrounds__ containing a reference to __backgrounds/Nature__ which, in turn has reference to {__Lighting__|__Perspective__} or something to that effect).

**Comment by adieyal** [comment]
That error means that the parser didn't understand your prompt, e.g. you didn't close a `}` somewhere. Does it ever happen with the built-in wildcards?

**Comment by thecakeisal1e** [comment]
So I replaced all my wildcards with the jumbo ones and did some generations with some random nested prompts to test if the stock wildcards worked, which they did (my nesting was pretty simple, though, and there weren't any wildcards down the tree that had conditionals that referenced the levels up). However, when I replaced all my wildcards and refreshed the manager, now _not a single gen_ with my previous prompt structure works and it throws "the same error"

> Error running process: .\stable-diffusion-webui\extensions\sd-dynamic-prompts\scripts\dynamic_prompting.py
Traceback (most recent call last):
  File ".\stable-diffusion-webui\modules\scripts.py", line 386, in process
    script.process(p, *script_args)
  File ".\stable-diffusion-webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py", line 413, in process
    all_prompts, all_negative_prompts = generate_prompts(
  File ".\stable-diffusion-webui\extensions\sd-dynamic-prompts\sd_dynamic_prompts\dynamic_prompting.py", line 76, in generate_prompts
    all_negative_prompts = negative_prompt_generator.generate(
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\generators\randomprompt.py", line 46, in generate
    return list(self._context.sample_prompts((template or ""), num_images))
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\sampling_context.py", line 97, in sample_prompts
    command = parse(prompt, parser_config=self.parser_config)
  File ".\stable-diffusion-webui\venv\lib\site-packages\dynamicprompts\parser\parse.py", line 302, in parse
    tokens = create_parser(parser_config=parser_config).parse_string(
  File ".\stable-diffusion-webui\venv\lib\site-packages\pyparsing\core.py", line 1141, in parse_string
    raise exc.with_traceback(None)
pyparsing.exceptions.ParseException: Expected end of text, found '{'  (at char 150), (line:1, col:151)

**Comment by thecakeisal1e** [comment]
So I rebuilt all my wildcards from scratch this weekend and I must've fixed some weird structure because I rarely get this issue (I still do but it isn't as pervasive. It'd be really nice to have some sort of debug option that can pinpoint exactly what wildcard or element is causing these errors in the future. Elsewise, this was my user error, it seems.

**Comment by adieyal** [comment]
Good to hear that it's resolved. I'll add the debug logs. If you are able to reliably reproduce the issue, please send me an example.

**Comment by NatureBeard** [comment]
This has happened to me before for two reasons, and most recently I discovered an easy way to troubleshoot and may help with better exception handling message debugging addition:

1) as mentioned it happens if you don't close {} or have an extra "{" when using the "|" choice syntax
2) if you call a wildcard and fail to use proper syntax. (i.e.  __subjects/cutimals__ but only put "subjects/cutimals__" or have an errant __ that isn't completed.)

The way I discovered which of my MANY wildcards was having the issue, I just deleted each one successively in my prompt and attempted to generate again, as error is raised immediately. If error was raised, I replaced that wildcard and deleted the next one in line until I found the one that was missing the syntax.
